src/main/java/com/aerospike/mapper/tools/LoadedObjectResolver.java

```diff
@@ -0,0 +1,42 @@
+package com.aerospike.mapper.tools;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.aerospike.client.Key;
+
+public class LoadedObjectResolver {
+	
+	private static class LoadedObjectMap {
+		private int referenceCount = 0;
```

Comment:
Why do we need the `referenceCount` variable? It's being only increased/decreased on begin/end sequentially?

Reply:
Correct, but the map only gets cleared when the reference count goes to 0. So say we have object A which references object B which references object A. Read() is called on A which will realize the reference to B, so read() is called on B. B then needs to load A, but the LoadedObjectResolver is called and returns the A from step 1 of the original read(). This prevents recursion. 

But when B finishes its read call, it should NOT clear the map. Otherwise, if A has a C which also has the same A, C might get an empty map which means it would re-read A and create a second (different) object reference to that A. The only time the map should be cleared is when the top level, original read is complete. Hence why the reference counting is needed.